spyder/api/plugins.py

```diff
@@ -304,115 +288,506 @@ def register_shortcut(self, qaction_or_qshortcut, context, name,
+        title = self.get_plugin_title()
+        if self.CONF_SECTION == 'editor':
+            title = _('Editor')
+
+        if self.shortcut is not None:
+            action = create_action(
+                self,
+                title,
+                toggled=lambda checked: self.toggle_view(checked),
+                shortcut=QKeySequence(self.shortcut),
+                context=Qt.WidgetShortcut,
+            )
+        else:
+            action = create_action(
+                self,
+                title,
+                toggled=lambda checked: self.toggle_view(checked),
+            )
+
+        self._toggle_view_action = action
+
+    @Slot()
+    def _close_window(self):
+        """Close QMainWindow instance that contains this plugin."""
+        if self._undocked_window is not None:
+            self._undocked_window.close()
+            self._undocked_window = None
+
+            # Oddly, these actions can appear disabled after the Dock
+            # action is pressed
+            self._undock_action.setDisabled(False)
+            self._close_plugin_action.setDisabled(False)
+
+    @Slot()
+    def _create_window(self):
+        """Create a QMainWindow instance containing this plugin."""
+        self._undocked_window = window = PluginWindow(self)
+        window.setAttribute(Qt.WA_DeleteOnClose)
+        icon = self.get_plugin_icon()
+
+        if is_text_string(icon):
+            icon = self.get_icon(icon)
+
+        window.setWindowIcon(icon)
+        window.setWindowTitle(self.get_plugin_title())
+        window.setCentralWidget(self)
+        window.resize(self.size())
+        self.refresh_plugin()
+        self.set_ancestor(window)
+        self.dockwidget.setFloating(False)
+        self.dockwidget.setVisible(False)
+
+        window.show()
+
+    # FIXME: Now this method sounds like a signal, not a method.
+    @Slot(bool)
+    def _on_top_level_changed(self, top_level):
+        """Actions to perform when a plugin is undocked to be moved."""
+        if top_level:
+            self._undock_action.setDisabled(True)
+        else:
+            self._undock_action.setDisabled(False)
+
+    # FIXME: Now this method sounds like a signal, not a method.
+    def _visibility_changed(self, enable):
+        """Dock widget visibility has changed."""
+        if self.dockwidget is None:
+            return
+
+        if enable:
+            self.dockwidget.raise_()
+            widget = self.get_focus_widget()
+
+            if widget is not None and self._undocked_window is not None:
+                widget.setFocus()
+
+        visible = self.dockwidget.isVisible() or self._ismaximized
+        if self.DISABLE_ACTIONS_WHEN_HIDDEN:
+            toggle_actions(self._plugin_actions, visible)
+
+        self._isvisible = enable and visible
+        if self._isvisible:
+            self.refresh_plugin()
+
+    # FIXME: This sounds like something that should be public?
```

Comment:
Agree, I think both the IPythonConsole and the Notebook plugin can use this to (de)activate menu items when appropriate. They also both use `self._options_menu` so perhaps that should be public as well.

Reply:
Yes